File: Sources/CCUsageCore/Engine/RouterPaths.swift
import Foundation

/// Onde tudo do produto mora no disco.
///
/// Um espaço só, em `Application Support`, com o `bundleID` como raiz. É estável
/// e independente do `CFBundleIdentifier` do `.app`, para a configuração
/// sobreviver a uma renomeação do binário.
public struct RouterPaths: Sendable {
    /// Namespace de armazenamento. Não muda quando o app é renomeado.
    ///
    /// **Diverge do `CFBundleIdentifier` de propósito, e não pode ser
    /// "corrigido".** Em 2026-09-18 o produto virou `falcao-token-router` e o
    /// bundle passou a ser `com.synqo.falcao-token-router`; este valor ficou
    /// como estava. O motivo é a credencial: a casa de cada conta é
    /// `<base>/accounts/<uuid>`, e o Claude Code nomeia o item de chaveiro do
    /// perfil como `Claude Code-credentials-<sha256(caminho)[:8]>`. Mexer na
    /// base muda o caminho, muda o hash, e TODA credencial fica inalcançável de
    /// uma vez — com os itens antigos órfãos no chaveiro e nenhuma mensagem que
    /// explique. Renomear aqui exige migrar diretórios e reescrever cada item de
    /// chaveiro antes de apagar o antigo; até que alguém queira pagar esse
    /// preço, o nome de pasta é histórico e fica.
    public static let bundleID = "com.synqo.falcao-router"

    public let base: URL

    public init(bundleID: String = RouterPaths.bundleID,
                appSupport: URL? = nil) {
        // `ROUTER_APP_SUPPORT` redireciona a raiz — usado em teste e para a CLI e
        // o app apontarem para o mesmo lugar sob outro id.
        let root = appSupport
            ?? ProcessInfo.processInfo.environment["ROUTER_APP_SUPPORT"].map(URL.init(fileURLWithPath:))
            ?? FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        self.base = root.appending(path: bundleID)
    }

    /// O `RouterConfig` inteiro, num JSON.
    public var configFile: URL { base.appending(path: "config.json") }

    /// A casa de cada conta — o perfil onde ela faz `auth login` e onde a
    /// credencial-mãe vive no chaveiro. Uma por conta, isolada.
    public func accountHome(_ id: UUID) -> ConfigDir {
        .dedicated(base.appending(path: "accounts").appending(path: id.uuidString).path)
    }

    /// O perfil onde as sessões de um grupo rodam. O grupo padrão usa `~/.claude`
    /// (sem variável de ambiente); os demais, um perfil dedicado.
    public func groupConfigDir(_ id: UUID, isDefault: Bool,
                               home: String = NSHomeDirectory()) -> ConfigDir {
        isDefault
            ? .standard(home: home)
            : .dedicated(base.appending(path: "groups").appending(path: id.uuidString).path)
    }

    /// Onde o sensor grava as amostras de uso.
    public var usageDir: URL {
        GroupUsageStore.directory(bundleID: RouterPaths.bundleID,
                                  base: base.deletingLastPathComponent())
    }
}

File: windows/crates/router-core/src/engine/router_paths.rs
//! Onde tudo do produto mora no disco.
//!
//! No Windows a raiz é `%LOCALAPPDATA%\com.synqo.falcao-router` (Local, não
//! Roaming — credencial não pode viajar entre máquinas). O namespace de dados
//! **diverge do bundle id de propósito e não pode ser "corrigido"** (é o mesmo do
//! macOS; mudá-lo tornaria toda credencial inalcançável de uma vez).

use std::env;
use std::path::PathBuf;

use super::config_dir::ConfigDir;
use super::group_usage::GroupUsageStore;
use crate::ids::Id;

#[derive(Clone, Debug)]
pub struct RouterPaths {
    pub base: PathBuf,
}

impl RouterPaths {
    /// Namespace de armazenamento. Não muda quando o app é renomeado.
    pub const BUNDLE_ID: &'static str = "com.synqo.falcao-router";

    /// Resolve a raiz: `ROUTER_APP_SUPPORT` (teste/override) → `%LOCALAPPDATA%`.
    pub fn new() -> Self {
        Self::with_app_support(None)
    }

    /// Raiz explícita (para teste), senão a resolução padrão.
    pub fn with_app_support(app_support: Option<PathBuf>) -> Self {
        let root = app_support
            .or_else(|| env::var_os("ROUTER_APP_SUPPORT").map(PathBuf::from))
            .or_else(|| env::var_os("LOCALAPPDATA").map(PathBuf::from))
            .unwrap_or_else(env::temp_dir);
        RouterPaths {
            base: root.join(Self::BUNDLE_ID),
        }
    }

    /// O `RouterConfig` inteiro, num JSON.
    pub fn config_file(&self) -> PathBuf {
        self.base.join("config.json")
    }

    /// Onde o sensor grava as amostras de uso: `<base>\usage`.
    pub fn usage_dir(&self) -> PathBuf {
        GroupUsageStore::directory(&self.base)
    }

    /// A casa de cada conta — o perfil onde ela faz `auth login`. Uma por conta.
    pub fn account_home(&self, id: Id) -> ConfigDir {
        ConfigDir::dedicated(
            self.base
                .join("accounts")
                .join(id.to_string())
                .to_string_lossy()
                .into_owned(),
        )
    }

    /// O perfil onde as sessões de um grupo rodam. O grupo padrão usa o
    /// `<home>\.claude` (sem variável); os demais, um perfil dedicado.
    pub fn group_config_dir(&self, id: Id, is_default: bool, home: &str) -> ConfigDir {
        if is_default {
            ConfigDir::standard(home)
        } else {
            ConfigDir::dedicated(
                self.base
                    .join("groups")
                    .join(id.to_string())
                    .to_string_lossy()
                    .into_owned(),
            )
        }
    }
}

impl Default for RouterPaths {
    fn default() -> Self {
        Self::new()
    }
}
